### ISSUE TITLE
Evolve Flag Incorrect Wandb Logging, Unreproducible Hyperparameter Performance

### ISSUE THREAD

**Issue by MannyGozzi** [ultralytics/yolov5#8567]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar bug report.

### Bug

The 10 epoch runs are using the following notebook script making use of the --evolve flag for hyperparameters.
!export WANDB_RUN_GROUP="evolution" && python ./deepsea-yolov5/yolov5/train.py \
--img=640 \
--data=./deepsea-yolov5/opt/ml/custom_config.yaml  \
--batch=2 \
--weights=yolov5s.pt \
--cfg=./deepsea-yolov5/yolov5/models/yolov5s.yaml \
--project="902005-vaa"\
--cache \
--epochs=10 \
--evolve=30
Evolution runs only output one point on the graph at the end of 10 epochs and the outputted hyperparameters do not show reproducible results when running in a 50 epoch run. The blue 50 epoch line showcases using the optimal hyperparameters which should intersect with the highest 10 epoch run, but it doesn't reach anywhere close. 
![image](https://user-images.githubusercontent.com/38353934/178777029-fce61b32-e4fc-4d36-8e6e-d38c7a4d18fb.png)

### Environment

-YOLOv5s
-Amazon Sagemaker
-Python 3.8

### Minimal Reproducible Example

# To run 30 evolutions at 10 epochs each
!export WANDB_RUN_GROUP="evolution" && python ./deepsea-yolov5/yolov5/train.py \
--img=640 \
--data=./deepsea-yolov5/opt/ml/custom_config.yaml  \
--batch=2 \
--weights=yolov5s.pt \
--cfg=./deepsea-yolov5/yolov5/models/yolov5s.yaml \
--project="902005-vaa"\
--cache \
--epochs=10 \
--evolve=30

# To run 50 epochs using optimal hyperparam
!export WANDB_RUN_GROUP="hyperparam" && python ./deepsea-yolov5/yolov5/train.py \
--img=640 \
--data=./deepsea-yolov5/opt/ml/custom_config.yaml  \
--batch=2 \
--weights=yolov5s.pt \
--cfg=./deepsea-yolov5/yolov5/models/yolov5s.yaml \
--hyp=./deepsea-yolov5/opt/ml/input/data/hyp.scratch-low.yaml \
--project="902005-vaa"\
--cache \
--epochs=50

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@MannyGozzi reproducibility is enabled for YOLOv5 only starting in torch==1.12.0 installs, but a 1.12.0 bug requires us to remove this option from requirements.txt, so 1.12.1 should be the first compatible torch release that allows YOLOv5 reproducibility.

--evolve is not intended for logging as only the final epoch of each generation is validated

**Comment by MannyGozzi** [comment]
> 

Ok but is there something that I'm doing wrong when running the 50 epoch run which results in totally different performance?

**Comment by glenn-jocher** [comment]
@MannyGozzi  👋 Hello! Thanks for asking about **training reproducibility**. A few percent difference in runs is not uncommon, i.e. multiple COCO trainings may result in +/- 0.5% mAP currently between runs.

YOLOv5 🚀 uses a single training seed which is set here using the `init_seeds()` function.
https://github.com/ultralytics/yolov5/blob/6865d19a92d8c160c7fc3c92256627dadce1cd1e/train.py#L108

This function sets `python`, `numpy` and `torch` seeds and updates `cudnn` settings:
https://github.com/ultralytics/yolov5/blob/6865d19a92d8c160c7fc3c92256627dadce1cd1e/utils/general.py#L122-L130

To set a new seed for example:
```python
    init_seeds(2 + RANK)  # set seed=2
```

Note that even when using the same seed trainings may produce different results, especially when using CUDA backends. See https://pytorch.org/docs/stable/notes/randomness.html for details on factors affecting PyTorch training reproducibility and sources of randomness.

<img width="949" alt="Screen Shot 2022-01-08 at 11 20 59 AM" src="https://user-images.githubusercontent.com/26833433/148660366-9155dfd9-351c-4174-818f-2b1b907b3e4e.png">

Good luck 🍀 and let us know if you have any other questions!

**Comment by MannyGozzi** [comment]
I've tried the proposed solution, however, the results still aren't as close as I'd like them to be. Is there a way to keep everything in the run the same? And is it possible to keep the anchors the same as well?

**Comment by glenn-jocher** [comment]
@MannyGozzi deterministic ops are enabled in YOLOv5 starting with torch 1.12
